lib/src/main/java/io/stormbird/token/tools/ParseMagicLink.java

```diff
@@ -92,81 +99,167 @@ public MagicLinkData parseUniversalLink(String link) throws SalesOrderMalformed
+        //now read signature
+        ds.readSignature(data.signature);
+        ds.close();
+        //now we have to build the message that the contract is expecting the signature for
+        data.message = getMessageFromCurrencyDropLink(
+                data.prefix,
+                nonceBytes,
+                amountBytes,
+                expiryBytes,
+                contractAddressBytes
+        );
+        return data;
+    }
+
+    //TODO make sure this isn't duplicate
+    private byte[] getSignatureFromComponents(byte v, byte[] r, byte[] s)
+    {
+        byte[] signature = new byte[65];
+        System.arraycopy(s, 0, signature, 0, 32);
+        System.arraycopy(r, 0, signature, 32, 32);
+        signature[64] = v;
+        return signature;
+    }
+
+    private MagicLinkData getMagicLinkDataFromURL(String linkData) throws SalesOrderMalformed
+    {
+        MagicLinkData data = new MagicLinkData();
+        byte[] fullOrder = cryptoInterface.Base64Decode(linkData);
         try
         {
             ByteArrayInputStream bas = new ByteArrayInputStream(fullOrder);
             EthereumReadBuffer ds = new EthereumReadBuffer(bas);
-
             data.contractType = ds.readByte();
-
             switch (data.contractType)
             {
-                case 0x00:
-                    data.contractType = 1;
-                    ds.reset();
-                    break;
+                case unassigned:
+                    return parseNonSpawnableLinks(linkData, false);
+                case normal:
+                    return parseNonSpawnableLinks(linkData, true);
+                case spawnable:
+                    return parseSpawnableLinks(linkData);
+                case customizable:
+                    //not yet implemented so default to spawnable
+                    return parseSpawnableLinks(linkData);
+                case currencyLink:
+                    return parseCurrencyLinks(linkData);
                 default:
-                    break;
+                    return parseNonSpawnableLinks(linkData, false);
             }
-
-            szabo = ds.toUnsignedLong(ds.readInt());
-            data.expiry = ds.toUnsignedLong(ds.readInt());
-            data.priceWei = Convert.toWei(BigDecimal.valueOf(szabo), Convert.Unit.SZABO).toBigInteger();
-            data.contractAddress = ds.readAddress();
-            //ticketCount = ds.available() / 2;
-            data.tickets = ds.readCompressedIndices(ds.available() - 65);
-            data.ticketCount = data.tickets.length;
-            //now read signature
-            ds.readSignature(data.signature);
-            ds.close();
-        } catch (IOException e) {
-            throw new SalesOrderMalformed();
-        } catch (StringIndexOutOfBoundsException f) {
-            throw new SalesOrderMalformed();
         } catch (Exception e) {
+            //TODO handle properly
             throw new SalesOrderMalformed();
```

Comment:
SalesOrderMalformed is caught in the classes that use this method. It's purposefully abstracted this way.

Reply:
Each method I have seen using this function only prints the stack trace... I should however move my comment to there instead so you're right.